File: Assets/Scripts/Networking/Protocol/WorldEntitySpawnedCodec.cs
using System;

namespace TopDownRoguelike.Networking.Protocol
{
    public static class WorldEntitySpawnedCodec
    {
        public const int PayloadSize =
            WorldStateSnapshotCodec.RecordSize;

        public static byte[] Encode(
            WorldEntityRecord record)
        {
            ValidateSpawnRecord(record);

            byte[] snapshotPayload =
                WorldStateSnapshotCodec.Encode(
                    new WorldStateSnapshotPayload(
                        new[] { record }));

            var payload =
                new byte[PayloadSize];

            Buffer.BlockCopy(
                snapshotPayload,
                WorldStateSnapshotCodec.PrefixSize,
                payload,
                0,
                PayloadSize);

            return payload;
        }

        public static WorldEntityRecord Decode(
            byte[] payload)
        {
            if (payload == null)
            {
                throw new ArgumentNullException(
                    nameof(payload));
            }

            if (payload.Length != PayloadSize)
            {
                throw new ArgumentException(
                    "World entity spawn payload must contain " +
                    "exactly one entity record.",
                    nameof(payload));
            }

            var snapshotPayload =
                new byte[
                    WorldStateSnapshotCodec.PrefixSize +
                    PayloadSize];

            PacketCodec.WriteNetworkUInt32(
                snapshotPayload,
                0,
                1u);

            Buffer.BlockCopy(
                payload,
                0,
                snapshotPayload,
                WorldStateSnapshotCodec.PrefixSize,
                PayloadSize);

            WorldEntityRecord record =
                WorldStateSnapshotCodec.Decode(
                    snapshotPayload)
                .Entities[0];

            ValidateSpawnRecord(record);
            return record;
        }

        private static void ValidateSpawnRecord(
            WorldEntityRecord record)
        {
            if (record == null)
            {
                throw new ArgumentNullException(
                    nameof(record));
            }

            if (record.Lifecycle !=
                WorldEntityLifecycle.Spawn)
            {
                throw new ArgumentException(
                    "World entity spawn record must use " +
                    "the Spawn lifecycle.",
                    nameof(record));
            }

            if ((record.Flags & WorldEntityFlags.Dead) != 0)
            {
                throw new ArgumentException(
                    "World entity spawn record cannot be dead.",
                    nameof(record));
            }
        }
    }
}

File: NetworkServer/src/protocol/WorldEntitySpawnedCodec.cpp
#include "protocol/WorldEntitySpawnedCodec.h"

#include <algorithm>
#include <stdexcept>

namespace
{
    void ValidateSpawnRecord(
        const tdr::protocol::WorldEntityRecord& record
    )
    {
        using namespace tdr::protocol;

        if (record.lifecycle !=
            WorldEntityLifecycle::Spawn)
        {
            throw std::invalid_argument(
                "World entity spawn record must use "
                "the Spawn lifecycle."
            );
        }

        if ((static_cast<std::uint16_t>(record.flags) &
            static_cast<std::uint16_t>(
                WorldEntityFlags::Dead)) != 0U)
        {
            throw std::invalid_argument(
                "World entity spawn record cannot be dead."
            );
        }
    }
}

namespace tdr::protocol
{
    std::vector<std::uint8_t>
        WorldEntitySpawnedCodec::Encode(
            const WorldEntityRecord& record
        )
    {
        ValidateSpawnRecord(record);

        WorldStateSnapshotPayload snapshot{};
        snapshot.entities.push_back(record);

        const auto snapshotPayload =
            WorldStateSnapshotCodec::Encode(snapshot);

        return std::vector<std::uint8_t>(
            snapshotPayload.begin() +
                static_cast<std::ptrdiff_t>(
                    kWorldStateSnapshotPrefixSize),
            snapshotPayload.end()
        );
    }

    WorldEntityRecord
        WorldEntitySpawnedCodec::Decode(
            const std::uint8_t* const data,
            const std::size_t size
        )
    {
        if (data == nullptr)
        {
            throw std::invalid_argument(
                "World entity spawn payload is null."
            );
        }

        if (size != kWorldEntitySpawnedPayloadSize)
        {
            throw std::invalid_argument(
                "World entity spawn payload must contain "
                "exactly one entity record."
            );
        }

        std::vector<std::uint8_t> snapshotPayload(
            kWorldStateSnapshotPrefixSize + size,
            0U
        );

        snapshotPayload[3] = 1U;

        std::copy(
            data,
            data + size,
            snapshotPayload.begin() +
                static_cast<std::ptrdiff_t>(
                    kWorldStateSnapshotPrefixSize)
        );

        const auto snapshot =
            WorldStateSnapshotCodec::Decode(
                snapshotPayload.data(),
                snapshotPayload.size());

        const auto record =
            snapshot.entities.front();

        ValidateSpawnRecord(record);
        return record;
    }
}
